### ISSUE TITLE
Bug in conversion to mps with non_blocking=True

### ISSUE THREAD

**Issue by Xuzzo** [pytorch/pytorch#139550]
### 🐛 Describe the bug

The following code fails:
```
import torch

a = torch.tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
a0 = a.clone().to("mps")

a = a.to("mps", non_blocking=True)
assert torch.equal(a0, a), f"{a}"
```
Resulting tensor has random values here and there, something like
```
tensor([               6,                1,                2,                3,
                       4,                5,                6,                7,
                       8, 1688849860263945], device='mps:0')
```

### Versions

PyTorch version: 2.5.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.7.1 (arm64)
GCC version: Could not collect
Clang version: 16.0.0 (clang-1600.0.26.4)
CMake version: version 3.30.5
Libc version: N/A

Python version: 3.10.15 (main, Oct  3 2024, 02:24:49) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-14.7.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Max

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.5.1
[pip3] torchvision==0.20.1
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] torch                     2.5.1                    pypi_0    pypi
[conda] torchvision               0.20.1                   pypi_0    pypi

**Comment by drisspg** [comment]
Does not reproduce on my machine on nightly

**Comment by janeyx99** [comment]
Relevant guide btw https://pytorch.org/tutorials/intermediate/pinmem_nonblock.html

**Comment by malfet** [comment]
@vmoens highlighted an even worse propblem: `mps.synchronize()` does not synchronize async copies 
```python
a = torch.arange(1000)
b = torch.arange(1000)
a = a.to('mps', non_blocking=True)
b = b.to('mps', non_blocking=True)
torch.mps.synchronize()
assert (a == b).all()
```

**Comment by qqaatw** [comment]
Offline conversations we had:

I managed to find the root cause of it - the commitAndContinue commits and pushes the all the encoded commands into rootCommandBuffer and flushes the current command buffer. That’s why the synchronize later doesn’t work because it actually commits on the brand new command buffer after flush, which doesn’t contain any commands.

Solution: use rootCommandBuffer instead or prevent using commitAndContinue, which possibly results in degraded perf.

Also this path is problematic cuz if we do two consequent MPSStream::commit()  the new command buffer would directly replace the previous _prevCommandBuffer, which won’t be waited if we subsequently call commitAndWait https://github.com/pytorch/pytorch/blob/891ba2ec8a3e2e71137fab4a8e91940a19c8272b/aten/src/ATen/mps/MPSStream.mm#L131-L133

**Comment by skotapati** [comment]
Shifting back to the original issue for a second, I actually wasn't able to get a repro using the script from @Xuzzo on a MacOS15 system. I was able to repro the issue on a MacOS14 system, but only by running it on loop using a script, and the error only occurred about 1% of the time.

@Xuzzo are you seeing this issue repro consistently, and would it be possible at all to upgrade to MacOS15+? I'll attempt to debug further using the repro cases I've been able to get so far

**Comment by Xuzzo** [comment]
@skotapati yes I see the issue consistently, almost all the times when I run the code on MacOS14. Will try upgrading to MacOS15.

**Comment by malfet** [comment]
@skotapati this almost always reproduces for me
```python
import torch

a = torch.arange(256, dtype=torch.uint8)
a = a.to("mps", non_blocking=True)
print(a[-10:])
```

**Comment by Xuzzo** [comment]
agree with @malfet that this could be considered expected behavior. Problem is that with cuda it works even in the first case and this generates quite a few bugs when switching from one device to another

**Comment by malfet** [comment]
@Xuzzo can you elaborate what works on CUDA? Doing something like `a = a.to("cuda", non_blocking=True)` might work, but not guaranteed to give you accurate results if no other reference to 'a' exists in the code.

**Comment by Xuzzo** [comment]
The same code, i.e.
```
import torch

a = torch.arange(256, dtype=torch.uint8)
a = a.to("cuda", non_blocking=True)
print(a[-10:])
```
works without issues for me. In fact, I first noticed the problem because a model trained on the gpu was not working well on my Mac (issue huggingface/lerobot#496)
